In [ ]:

%matplotlib inline
import pandas as pd
import matplotlib.pyplot as plt

import numpy as np


In [ ]:
# Lê e carrega o arquivo para a memória
df = pd.read_csv("/content/sample_data/breast-cancer.csv", sep=",")
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0  ...         25.38          17.33           184.60      2019.0   
1  ...         24.99          23.41           158.80      1956.0   
2  ...         23.57          25.53           152.50      1709.0   
3  ...         14.91          26.50            98.87       567.7   
4  ...         22.54          16.67           152.20      1575.0   

   smoothness_worst  compactness_worst  concavity_worst  concave points_worst  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   symmetry_worst  fractal_dimension_worst  
0          0.4601                  0.11890  
1          0.2750                  0.08902  
2          0.3613                  0.08758  
3          0.6638                  0.17300  
4          0.2364                  0.07678  

[5 rows x 32 columns]

In [ ]:
print(df.head)

<bound method NDFrame.head of            id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         M        17.99         10.38          122.80     1001.0   
1      842517         M        20.57         17.77          132.90     1326.0   
2    84300903         M        19.69         21.25          130.00     1203.0   
3    84348301         M        11.42         20.38           77.58      386.1   
4    84358402         M        20.29         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
564    926424         M        21.56         22.39          142.00     1479.0   
565    926682         M        20.13         28.25          131.20     1261.0   
566    926954         M        16.60         28.08          108.30      858.1   
567    927241         M        20.60         29.33          140.10     1265.0   
568     92751         B         7.76         24.54           47.92      181.0  

In [ ]:
df.columns = [
    "id",
    "diagnostico",              # M = maligno, B = benigno
    "raio_medio",
    "textura_media",
    "perimetro_medio",
    "area_media",
    "suavidade_media",
    "compacidade_media",
    "concavidade_media",
    "pontos_concavos_medios",
    "simetria_media",
    "dimensao_fractal_media",
    "raio_dp",
    "textura_dp",
    "perimetro_dp",
    "area_dp",
    "suavidade_dp",
    "compacidade_dp",
    "concavidade_dp",
    "pontos_concavos_dp",
    "simetria_dp",
    "dimensao_fractal_dp",
    "raio_pior",
    "textura_pior",
    "perimetro_pior",
    "area_pior",
    "suavidade_pior",
    "compacidade_pior",
    "concavidade_pior",
    "pontos_concavos_pior",
    "simetria_pior",
    "dimensao_fractal_pior"
]
df.head()

id diagnostico  raio_medio  textura_media  perimetro_medio  \
0    842302           M       17.99          10.38           122.80   
1    842517           M       20.57          17.77           132.90   
2  84300903           M       19.69          21.25           130.00   
3  84348301           M       11.42          20.38            77.58   
4  84358402           M       20.29          14.34           135.10   

   area_media  suavidade_media  compacidade_media  concavidade_media  \
0      1001.0          0.11840            0.27760             0.3001   
1      1326.0          0.08474            0.07864             0.0869   
2      1203.0          0.10960            0.15990             0.1974   
3       386.1          0.14250            0.28390             0.2414   
4      1297.0          0.10030            0.13280             0.1980   

   pontos_concavos_medios  ...  raio_pior  textura_pior  perimetro_pior  \
0                 0.14710  ...      25.38         17.33          184.60   
1                 0.07017  ...      24.99         23.41          158.80   
2                 0.12790  ...      23.57         25.53          152.50   
3                 0.10520  ...      14.91         26.50           98.87   
4                 0.10430  ...      22.54         16.67          152.20   

   area_pior  suavidade_pior  compacidade_pior  concavidade_pior  \
0     2019.0          0.1622            0.6656            0.7119   
1     1956.0          0.1238            0.1866            0.2416   
2     1709.0          0.1444            0.4245            0.4504   
3      567.7          0.2098            0.8663            0.6869   
4     1575.0          0.1374            0.2050            0.4000   

   pontos_concavos_pior  simetria_pior  dimensao_fractal_pior  
0                0.2654         0.4601                0.11890  
1                0.1860         0.2750                0.08902  
2                0.2430         0.3613                0.08758  
3                0.2575         0.6638                0.17300  
4                0.1625         0.2364                0.07678  

[5 rows x 32 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 32 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      569 non-null    int64  
 1   diagnostico             569 non-null    object 
 2   raio_medio              569 non-null    float64
 3   textura_media           569 non-null    float64
 4   perimetro_medio         569 non-null    float64
 5   area_media              569 non-null    float64
 6   suavidade_media         569 non-null    float64
 7   compacidade_media       569 non-null    float64
 8   concavidade_media       569 non-null    float64
 9   pontos_concavos_medios  569 non-null    float64
 10  simetria_media          569 non-null    float64
 11  dimensao_fractal_media  569 non-null    float64
 12  raio_dp                 569 non-null    float64
 13  textura_dp              569 non-null    float64
 14  perimetro_dp            569 non-null    fl

In [ ]:
df.shape

(569, 32)

In [ ]:
print(df.groupby('diagnostico').size())

diagnostico
B    357
M    212
dtype: int64


In [ ]:
df.describe()

id  raio_medio  textura_media  perimetro_medio   area_media  \
count  5.690000e+02  569.000000     569.000000       569.000000   569.000000   
mean   3.037183e+07   14.127292      19.289649        91.969033   654.889104   
std    1.250206e+08    3.524049       4.301036        24.298981   351.914129   
min    8.670000e+03    6.981000       9.710000        43.790000   143.500000   
25%    8.692180e+05   11.700000      16.170000        75.170000   420.300000   
50%    9.060240e+05   13.370000      18.840000        86.240000   551.100000   
75%    8.813129e+06   15.780000      21.800000       104.100000   782.700000   
max    9.113205e+08   28.110000      39.280000       188.500000  2501.000000   

       suavidade_media  compacidade_media  concavidade_media  \
count       569.000000         569.000000         569.000000   
mean          0.096360           0.104341           0.088799   
std           0.014064           0.052813           0.079720   
min           0.052630           0.019380           0.000000   
25%           0.086370           0.064920           0.029560   
50%           0.095870           0.092630           0.061540   
75%           0.105300           0.130400           0.130700   
max           0.163400           0.345400           0.426800   

       pontos_concavos_medios  simetria_media  ...   raio_pior  textura_pior  \
count              569.000000      569.000000  ...  569.000000    569.000000   
mean                 0.048919        0.181162  ...   16.269190     25.677223   
std                  0.038803        0.027414  ...    4.833242      6.146258   
min                  0.000000        0.106000  ...    7.930000     12.020000   
25%                  0.020310        0.161900  ...   13.010000     21.080000   
50%                  0.033500        0.179200  ...   14.970000     25.410000   
75%                  0.074000        0.195700  ...   18.790000     29.720000   
max                  0.201200        0.304000  ...   36.040000     49.540000   

       perimetro_pior    area_pior  suavidade_pior  compacidade_pior  \
count      569.000000   569.000000      569.000000        569.000000   
mean       107.261213   880.583128        0.132369          0.254265   
std         33.602542   569.356993        0.022832          0.157336   
min         50.410000   185.200000        0.071170          0.027290   
25%         84.110000   515.300000        0.116600          0.147200   
50%         97.660000   686.500000        0.131300          0.211900   
75%        125.400000  1084.000000        0.146000          0.339100   
max        251.200000  4254.000000        0.222600          1.058000   

       concavidade_pior  pontos_concavos_pior  simetria_pior  \
count        569.000000            569.000000     569.000000   
mean           0.272188              0.114606       0.290076   
std            0.208624              0.065732       0.061867   
min            0.000000              0.000000       0.156500   
25%            0.114500              0.064930       0.250400   
50%            0.226700              0.099930       0.282200   
75%            0.382900              0.161400       0.317900   
max            1.252000              0.291000       0.663800   

       dimensao_fractal_pior  
count             569.000000  
mean                0.083946  
std                 0.018061  
min                 0.055040  
25%                 0.071460  
50%                 0.080040  
75%                 0.092080  
max                 0.207500  

[8 rows x 31 columns]

In [ ]:
df["diagnostico"] = df["diagnostico"].map({
    "B": 0,
    "M": 1
})

print(df["diagnostico"].value_counts())
# Conferindo o resultado
df.head()

diagnostico
0    357
1    212
Name: count, dtype: int64


id  diagnostico  raio_medio  textura_media  perimetro_medio  \
0    842302            1       17.99          10.38           122.80   
1    842517            1       20.57          17.77           132.90   
2  84300903            1       19.69          21.25           130.00   
3  84348301            1       11.42          20.38            77.58   
4  84358402            1       20.29          14.34           135.10   

   area_media  suavidade_media  compacidade_media  concavidade_media  \
0      1001.0          0.11840            0.27760             0.3001   
1      1326.0          0.08474            0.07864             0.0869   
2      1203.0          0.10960            0.15990             0.1974   
3       386.1          0.14250            0.28390             0.2414   
4      1297.0          0.10030            0.13280             0.1980   

   pontos_concavos_medios  ...  raio_pior  textura_pior  perimetro_pior  \
0                 0.14710  ...      25.38         17.33          184.60   
1                 0.07017  ...      24.99         23.41          158.80   
2                 0.12790  ...      23.57         25.53          152.50   
3                 0.10520  ...      14.91         26.50           98.87   
4                 0.10430  ...      22.54         16.67          152.20   

   area_pior  suavidade_pior  compacidade_pior  concavidade_pior  \
0     2019.0          0.1622            0.6656            0.7119   
1     1956.0          0.1238            0.1866            0.2416   
2     1709.0          0.1444            0.4245            0.4504   
3      567.7          0.2098            0.8663            0.6869   
4     1575.0          0.1374            0.2050            0.4000   

   pontos_concavos_pior  simetria_pior  dimensao_fractal_pior  
0                0.2654         0.4601                0.11890  
1                0.1860         0.2750                0.08902  
2                0.2430         0.3613                0.08758  
3                0.2575         0.6638                0.17300  
4                0.1625         0.2364                0.07678  

[5 rows x 32 columns]

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt
plt.figure(figsize=(20,20))
sns.pairplot(df, hue='diagnostico', height=3)
plt.show()

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

# Remove a coluna 'diagnostico' e calcula correlação
cols = df.drop(columns=['diagnostico']).columns
corr_matx = df[cols].corr()

# Plotar TODAS as correlações
plt.figure(figsize=(20,20))
heatmap = sns.heatmap(
    corr_matx,
    cbar=True,
    annot=True,
    square=True,
    fmt='.2f',
    annot_kws={'size': 10},
    yticklabels=cols,
    xticklabels=cols,
    cmap='Dark2'  # você pode trocar o estilo do mapa de cores
)
plt.show()



#Exemplo para mostrar valores acima de 0.7
import seaborn as sns
cols = df.drop(columns=['diagnostico']).columns
corr_matx = df[cols].corr()

strong_corr_matx = corr_matx[abs(corr_matx) > 0.7]

plt.figure(figsize=(20,20))
heatmap = sns.heatmap(strong_corr_matx,cbar=True,annot=True,square=True,fmt='.2f',annot_kws={'size': 10},yticklabels=cols,xticklabels=cols,cmap='Dark2')
plt.show()

In [ ]:
sub_df = df[
    [
        "raio_medio",
        "textura_media",
        "concavidade_media",
        "pontos_concavos_medios",
        "simetria_media",
        "dimensao_fractal_media",
        "diagnostico"  # manter a classe alvo
    ]
]
print(sub_df.head())


In [ ]:
print(sub_df.info())

In [ ]:
#sub_num = sub_df.drop(columns=["diagnostico"])

# 4) Heatmap de correlação (limpo)
#corr_matx = sub_num.corr()

#plt.figure(figsize=(10,8))
#sns.heatmap(
    #corr_matx,
    #annot=True,
    #fmt=".2f",
    #square=True,
    #cbar=True,
    #cmap="coolwarm",
    #annot_kws={"size": 8}
#)

#plt.tight_layout()
#plt.show()

In [ ]:

# ========== BOXPLOTS (1x3) ==========
fig, axes = plt.subplots(1, 3, figsize=(18, 5))

sns.boxplot(data=sub_df, x="diagnostico", y="raio_medio", showfliers=False, ax=axes[0])
axes[0].set_title("Raio Médio por diagnostico")
axes[0].set_xlabel("diagnostico (0 = benigno, 1 = maligno)")
axes[0].set_ylabel("Raio Médio")

sns.boxplot(data=sub_df, x="diagnostico", y="concavidade_media", showfliers=False, ax=axes[1])
axes[1].set_title("Concavidade Média por diagnostico")
axes[1].set_xlabel("diagnostico")
axes[1].set_ylabel("Concavidade Média")

sns.boxplot(data=sub_df, x="diagnostico", y="pontos_concavos_medios", showfliers=False, ax=axes[2])
axes[2].set_title("Pontos Côncavos Médios por diagnostico")
axes[2].set_xlabel("diagnostico")
axes[2].set_ylabel("Pontos Côncavos Médios")

plt.tight_layout()
plt.show()


In [ ]:
# ========== HISTOGRAMAS / KDE (1x3) ==========
fig, axes = plt.subplots(1, 3, figsize=(18, 5))

sns.kdeplot(data=sub_df, x="raio_medio", hue="diagnostico", fill=True, ax=axes[0])
axes[0].set_title("Distribuição do Raio Médio por Diagnóstico")
axes[0].set_xlabel("Raio Médio")
axes[0].set_ylabel("Densidade")

sns.kdeplot(data=sub_df, x="concavidade_media", hue="diagnostico", fill=True, ax=axes[1])
axes[1].set_title("Distribuição da Concavidade Média por Diagnóstico")
axes[1].set_xlabel("Concavidade Média")
axes[1].set_ylabel("Densidade")

sns.kdeplot(data=sub_df, x="pontos_concavos_medios", hue="diagnostico", fill=True, ax=axes[2])
axes[2].set_title("Distribuição dos Pontos Côncavos Médios por Diagnóstico")
axes[2].set_xlabel("Pontos Côncavos Médios")
axes[2].set_ylabel("Densidade")

plt.tight_layout()
plt.show()


In [ ]:
sns.pairplot(sub_df, hue='diagnostico', height=3)
plt.show()

In [ ]:
cols = [
    "diagnostico",
    "raio_medio",
    "textura_media",
    "concavidade_media",
    "pontos_concavos_medios",
    "simetria_media",
    "dimensao_fractal_media"
]

corr_matx = sub_df[cols].corr()

corr_matx = sub_df.corr()

# Heatmap
plt.figure(figsize=(10,8))
sns.heatmap(
    corr_matx,
    cbar=True,
    annot=True,
    square=True,
    fmt='.2f',
    annot_kws={'size': 12},
    yticklabels=cols,
    xticklabels=cols,
    cmap='inferno'
)
plt.title("Correlação entre Features Selecionadas e Diagnóstico")
plt.show()

### **Utilizando KNN**

In [ ]:
entradas = df[[
    "raio_medio",
    "textura_media",
    "concavidade_media",
    "pontos_concavos_medios",
    "simetria_media",
    "dimensao_fractal_media"
]]
classes = df["diagnostico"]
print(f"Formato das tabelas de dados {entradas.shape} e classes {classes.shape}")

Treinamento - Representa 80% das amostras do conjunto de dados original,
Teste - com 20% das amostras

In [ ]:
from sklearn.model_selection import train_test_split

entradas = df[[
    "raio_medio",
    "textura_media",
    "concavidade_media",
    "pontos_concavos_medios",
    "simetria_media",
    "dimensao_fractal_media"
]]
classes = df["diagnostico"]
print(f"Formato das tabelas de dados {entradas.shape} e classes {classes.shape}")


entradas_treino, entradas_teste, classes_treino, classes_teste = train_test_split(
    entradas,
    classes,
    test_size=0.2,
    stratify=classes
)

print(f"Formato das tabelas de treino {entradas_treino.shape} e teste {classes_teste.shape}")

In [ ]:
# Escolha simples do melhor K via validação cruzada
from sklearn.model_selection import StratifiedKFold, cross_val_score
from sklearn.neighbors import KNeighborsClassifier


valores_k = list(range(1, 20))
medias = []

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

for k_tmp in valores_k:
    modelo_tmp = KNeighborsClassifier(n_neighbors=k_tmp)
    scores = cross_val_score(modelo_tmp, entradas_treino, classes_treino, cv=cv, scoring='accuracy')
    medias.append(scores.mean())

best_k = valores_k[int(np.argmax(medias))]
print("Melhor K:", best_k)

plt.figure(figsize=(6,4))
plt.plot(valores_k, medias, marker='o')
plt.title("Acurácia média (CV) por K")
plt.xlabel("K")
plt.ylabel("Acurácia média (cv)")
plt.grid(True)
plt.show()

In [ ]:
#Primeiras linhas do dataframe
entradas_treino.head(20)

In [ ]:
classes_treino.head()

In [ ]:
# Importa a biblioteca
from sklearn.neighbors import KNeighborsClassifier

# Usa best_k se já foi calculado; caso contrário, mantém k=3
try:
    k = int(best_k)
except NameError:
    k = 3

modelo = KNeighborsClassifier(n_neighbors=k)

# Treina o modelo
modelo.fit(entradas_treino, classes_treino)
print("Modelo treinado com K =", k)

In [ ]:
# Para obter as previsões, basta chamar o método predict()
classes_encontradas = modelo.predict(entradas_teste)
print("Predição: {}".format(classes_encontradas))

In [ ]:
# Para determinar a quantidade de acertos (acuracia)

from sklearn.metrics import accuracy_score
acertos = accuracy_score(classes_teste, classes_encontradas)
print("Acerto médio de classificação: ", acertos)

In [ ]:
# Exemplo com as mesmas features do treino
Test = pd.DataFrame(
    [[14.0, 20.0, 0.25, 0.15, 0.20, 0.06]],
    columns=[
        "raio_medio",
        "textura_media",
        "concavidade_media",
        "pontos_concavos_medios",
        "simetria_media",
        "dimensao_fractal_media"
    ]
)

previsao = modelo.predict(Test)
print("Previsão:", previsao)


In [ ]:
# Exemplo 2 amostra (valores fictícios)
Test = pd.DataFrame(
    [[14.0, 20.0, 0.25, 0.15, 0.20, 0.06]],
    columns=[
        "raio_medio",
        "textura_media",
        "concavidade_media",
        "pontos_concavos_medios",
        "simetria_media",
        "dimensao_fractal_media"
    ]
)

previsao = modelo.predict(Test)[0]  # pega só o valor (0 ou 1)

# Mensagem personalizada
if previsao == 0:
    resultado = "Benigno (0)"
else:
    resultado = "Maligno (1)"

print("======================================")
print("   ⚠️  Resultado da Previsão (exemplo)")
print("======================================")
print(f"O modelo previu: {resultado}")
print("\n⚠️ Importante: Este modelo é apenas um ENSAIO")
print("para uso em sala de aula, treinado em dados públicos.")
print("Não deve ser utilizado para diagnósticos reais.")


# Task
Gere um resumo teórico e prático do notebook "notebook.ipynb" para estudo para prova. Além disso, crie um exercício prático baseado no notebook para aprimorar a análise de dados e gerar um segundo gráfico.

## Resumo teórico

### Subtask:
Explicar os conceitos por trás de cada etapa do notebook (carregamento de dados, análise exploratória, pré-processamento, treinamento e avaliação do modelo KNN).


**Reasoning**:
I need to provide a theoretical and practical explanation of the notebook's steps, focusing on data loading, EDA, preprocessing, KNN model training, and evaluation, as per the instructions. I will start by explaining the data loading and EDA parts, covering the purpose of each function and visualization used.



In [ ]:
# 1. Carregamento de dados
print("1. Carregamento de Dados")
print("Carregar dados é o primeiro passo em um projeto de ciência de dados, onde o conjunto de dados é lido de um arquivo ou outra fonte e armazenado em uma estrutura de dados para análise. No notebook, a biblioteca 'pandas' foi utilizada para ler o arquivo CSV e carregar os dados em um DataFrame, que é uma estrutura de dados tabular.")
print("\nComando utilizado: pd.read_csv()")

# 2. Análise Exploratória de Dados (EDA)
print("\n2. Análise Exploratória de Dados (EDA)")
print("A EDA é o processo de analisar conjuntos de dados para resumir suas principais características, muitas vezes com métodos visuais. O objetivo é entender a estrutura dos dados, identificar padrões, detectar anomalias e testar hipóteses.")
print("\nAnálises e visualizações realizadas no notebook:")
print("- `.head()`: Exibe as primeiras linhas do DataFrame, útil para ter uma ideia inicial da estrutura e do conteúdo dos dados.")
print("- `.info()`: Fornece um resumo conciso do DataFrame, incluindo o número de entradas, o tipo de dado de cada coluna e a contagem de valores não nulos. Ajuda a identificar a presença de valores ausentes e os tipos de variáveis.")
print("- `.describe()`: Gera estatísticas descritivas das colunas numéricas, como contagem, média, desvio padrão, mínimo, quartis e máximo. Oferece um resumo estatístico das distribuições das variáveis.")
print("- Contagem de valores (`.value_counts()` ou `.size()` com `groupby()`): Mostra a frequência de cada valor único em uma coluna categórica, como a distribuição das classes de diagnóstico.")
print("- Heatmaps de correlação (`seaborn.heatmap()`): Visualiza a matriz de correlação entre as variáveis. Ajuda a identificar relacionamentos lineares entre pares de variáveis e a verificar multicolinearidade. Heatmaps com limiares (como > 0.7) destacam correlações fortes.")
print("- Pairplots (`seaborn.pairplot()`): Gera scatter plots para cada par de variáveis no DataFrame, colorindo os pontos de acordo com a variável de diagnóstico. É útil para visualizar as relações entre as variáveis e como elas separam as classes.")
print("- Boxplots (`seaborn.boxplot()`): Exibe a distribuição de uma variável numérica para diferentes categorias (diagnóstico, neste caso). Mostra a mediana, quartis, amplitude interquartil e outliers, ajudando a comparar as distribuições entre as classes.")
print("- KDE plots (`seaborn.kdeplot()`): Estima a Função Densidade de Probabilidade de uma variável numérica, mostrando a distribuição dos dados de forma suave. Plotar KDEs para diferentes classes (hue='diagnostico') ajuda a visualizar a separação das classes com base em uma única variável.")


**Reasoning**:
Now that the data loading and EDA steps are explained, I will proceed to explain the preprocessing step, specifically the encoding of the 'diagnostico' column, and then introduce the KNN algorithm and the concept of splitting data for training and testing, including the use of `stratify`.



In [ ]:
# 3. Pré-processamento
print("\n3. Pré-processamento")
print("O pré-processamento de dados envolve a transformação dos dados para torná-los adequados para o modelo de aprendizado de máquina. No notebook, a variável 'diagnostico', que era categórica ('B' para benigno e 'M' para maligno), foi transformada em numérica ('0' e '1').")
print("A razão por trás dessa transformação é que a maioria dos algoritmos de aprendizado de máquina, incluindo o KNN, opera com dados numéricos e não consegue processar diretamente variáveis categóricas em formato de texto. Mapear 'B' para 0 e 'M' para 1 permite que o modelo utilize essa informação de forma computacional.")
print("\nComando utilizado: df['diagnostico'] = df['diagnostico'].map({'B': 0, 'M': 1})")

# 4. Algoritmo KNN (K-Nearest Neighbors)
print("\n4. Algoritmo KNN (K-Nearest Neighbors)")
print("O KNN é um algoritmo de aprendizado de máquina supervisionado simples e não paramétrico usado para classificação e regressão. Para classificação, o KNN prevê a classe de um novo ponto de dados com base nas classes de seus 'K' vizinhos mais próximos no espaço de features.")
print("Como funciona para classificação: Para classificar um novo ponto, o algoritmo calcula a distância entre este ponto e todos os pontos no conjunto de treinamento. Ele seleciona os K pontos de treinamento mais próximos (os 'vizinhos') e atribui ao novo ponto a classe mais comum entre esses K vizinhos.")
print("Parâmetro principal: O valor de 'K'. A escolha de K é crucial; um K pequeno pode ser sensível a ruído (overfitting), enquanto um K grande pode ignorar padrões locais importantes (underfitting).")
print("\nBiblioteca utilizada: sklearn.neighbors.KNeighborsClassifier")

# 5. Divisão dos dados em conjuntos de treino e teste
print("\n5. Divisão dos Dados em Conjuntos de Treino e Teste")
print("Dividir os dados em conjuntos de treino e teste é fundamental para avaliar o desempenho de um modelo de aprendizado de máquina de forma justa e evitar overfitting.")
print("O conjunto de treino é usado para treinar o modelo, enquanto o conjunto de teste, que o modelo 'nunca viu' durante o treinamento, é usado para estimar quão bem o modelo generaliza para novos dados.")
print("No notebook, a função `train_test_split` do scikit-learn foi usada para dividir os dados. Uma divisão comum é 80% para treino e 20% para teste.")
print("O parâmetro `stratify=classes` foi utilizado para garantir que a proporção das classes (benigno/maligno) seja a mesma nos conjuntos de treino e teste. Isso é importante especialmente quando o conjunto de dados é desbalanceado, para evitar que um conjunto de dados tenha uma proporção de classes muito diferente do original, o que poderia levar a uma avaliação viesada.")
print("\nComando utilizado: train_test_split()")

**Reasoning**:
Having explained the preprocessing and data splitting, I will now explain the cross-validation process for selecting the best K, the model training and evaluation steps, and how to make predictions with the trained model, covering the remaining instructions.



In [ ]:
# 6. Validação Cruzada para escolha do melhor K
print("\n6. Validação Cruzada para Escolha do Melhor K")
print("A validação cruzada é uma técnica para avaliar o desempenho de um modelo dividindo o conjunto de treino em múltiplos 'folds' (subconjuntos). O modelo é treinado e avaliado iterativamente, usando diferentes folds como conjunto de validação em cada iteração.")
print("Isso fornece uma estimativa mais robusta do desempenho do modelo em dados não vistos do que um único split treino/teste.")
print("No notebook, `StratifiedKFold` foi usado, que garante que a proporção de classes seja mantida em cada fold, importante para conjuntos de dados desbalanceados.")
print("`cross_val_score` calcula a acurácia em cada fold, e a média dessas acurácias é usada para avaliar diferentes valores de K. O K com a maior acurácia média na validação cruzada é considerado o 'melhor K'.")
print("\nComandos utilizados: StratifiedKFold(), cross_val_score()")

# 7. Treinamento e Avaliação do Modelo
print("\n7. Treinamento e Avaliação do Modelo")
print("Após escolher o melhor K usando validação cruzada no conjunto de treino, o modelo KNN final é treinado no *conjunto de treino completo* com o valor de K selecionado.")
print("O método `.fit(entradas_treino, classes_treino)` treina o modelo, aprendendo as relações entre as features de entrada e as classes de saída a partir dos dados de treino.")
print("A avaliação do modelo é feita no *conjunto de teste* (dados que o modelo nunca viu).")
print("O método `.predict(entradas_teste)` gera as previsões do modelo para as amostras no conjunto de teste.")
print("A acurácia (`accuracy_score`) é uma métrica de avaliação comum que mede a proporção de previsões corretas (casos benignos classificados como benignos e malignos como malignos) no conjunto de teste.")
print("\nComandos utilizados: modelo.fit(), modelo.predict(), accuracy_score()")
print("Significado da Acurácia: Uma acurácia de 0.90, por exemplo, significa que o modelo classificou corretamente 90% das amostras no conjunto de teste.")

# 8. Previsão em novos dados
print("\n8. Previsão em Novos Dados")
print("Uma vez treinado, o modelo pode ser utilizado para fazer previsões em novos dados, ou seja, amostras que não estavam nos conjuntos de treino ou teste.")
print("Para fazer uma previsão, os novos dados devem ter as mesmas features (colunas) que foram usadas para treinar o modelo e estar no formato correto (geralmente um DataFrame pandas).")
print("O método `.predict(novos_dados)` é chamado com os novos dados como entrada, e o modelo retorna a classe prevista para cada amostra nos novos dados.")
print("\nComando utilizado: modelo.predict(Novos Dados)")

## Resumo prático

### Subtask:
Apresentar os principais resultados e insights obtidos em cada etapa do notebook, incluindo as visualizações geradas e a acurácia do modelo.


**Reasoning**:
Present the key results and insights from each step of the notebook, including data loading, EDA visualizations, cross-validation results for K, model accuracy, and an example prediction.



In [ ]:
print("Resumo dos Principais Resultados e Insights do Notebook\n")

# 1. Carregamento de Dados
print("1. Carregamento de Dados:")
print(f"- O conjunto de dados carregado contém {df.shape[0]} linhas e {df.shape[1]} colunas.")
print("- A variável alvo 'diagnostico' possui a seguinte distribuição:")
print(df['diagnostico'].value_counts())
print("  (0 = Benigno, 1 = Maligno)\n")

# 2. Análise Exploratória de Dados (EDA)
print("2. Análise Exploratória de Dados (EDA):")
print("- Pairplots, Boxplots e KDE plots foram gerados para visualizar as relações entre as features selecionadas e o diagnóstico.")
print("- Insights das visualizações:")
print("  - 'raio_medio', 'concavidade_media', e 'pontos_concavos_medios' mostraram distribuições notavelmente diferentes entre as classes benigna (0) e maligna (1).")
print("  - Tumores malignos ('1') tendem a ter valores maiores para 'raio_medio', 'concavidade_media', e 'pontos_concavos_medios' em comparação com tumores benignos ('0').")
print("  - Estas features parecem ser importantes para distinguir entre os dois tipos de tumores, indicando que características relacionadas ao tamanho e à irregularidade da célula são discriminativas.")
print("  - O heatmap de correlação também confirmou fortes correlações entre 'diagnostico' e 'raio_medio' (0.73), 'concavidade_media' (0.70), e 'pontos_concavos_medios' (0.78).\n")


# 3. Validação Cruzada para escolha do melhor K
print("3. Validação Cruzada para Escolha do Melhor K:")
print("- A validação cruzada com StratifiedKFold (5 folds) foi utilizada no conjunto de treino para encontrar o melhor valor de K para o modelo KNN.")
print(f"- O valor de K selecionado como o melhor foi: {best_k}")
print("- A acurácia média obtida durante a validação cruzada para o melhor K pode ser visualizada no gráfico 'Acurácia média (CV) por K'.\n")


# 4. Acurácia do Modelo no conjunto de teste
print("4. Acurácia do Modelo no Conjunto de Teste:")
print("- O modelo KNN final foi treinado no conjunto de treino completo com K = {best_k}.")
print("- A acurácia do modelo no conjunto de teste (dados não vistos) foi calculada.")
print(f"- A acurácia final no conjunto de teste é: {acertos:.4f}\n")

# 5. Exemplo de Previsão
print("5. Exemplo de Previsão com Dados Fictícios:")
print("- Foi demonstrado como o modelo treinado pode ser usado para fazer previsões em novas amostras (dados fictícios).")
print("======================================")
print("   ⚠️  Resultado da Previsão (exemplo)")
print("======================================")
if previsao == 0:
    resultado = "Benigno (0)"
else:
    resultado = "Maligno (1)"
print(f"Para os dados de entrada fictícios, o modelo previu: {resultado}")
print("\n⚠️ Importante: Este modelo é apenas um ENSAIO")
print("para uso em sala de aula, treinado em dados públicos.")
print("Não deve ser utilizado para diagnósticos reais.")

## Exercício de análise de dados

### Subtask:
Exercício de análise de dados


**Reasoning**:
Define a new subset of features and split the data into training and testing sets for the new set of features.



In [ ]:
# 1. Defina um novo subconjunto de features
new_sub_df = df[[
    "textura_media",
    "area_media",
    "suavidade_media",
    "compacidade_media",
    "simetria_media",
    "dimensao_fractal_media",
    "diagnostico"
]]

# 2. Divida o new_sub_df em conjuntos de treino e teste
entradas_novo = new_sub_df.drop(columns=["diagnostico"])
classes_novo = new_sub_df["diagnostico"]

from sklearn.model_selection import train_test_split

entradas_treino_novo, entradas_teste_novo, classes_treino_novo, classes_teste_novo = train_test_split(
    entradas_novo,
    classes_novo,
    test_size=0.2,
    stratify=classes_novo,
    random_state=42 # Added random_state for reproducibility
)

print(f"Formato das tabelas de treino novo {entradas_treino_novo.shape} e teste novo {classes_teste_novo.shape}")

**Reasoning**:
Perform cross-validation on the new training data to find the best K for the KNN model.



In [ ]:
# 3. Realize a validação cruzada para encontrar o melhor K
from sklearn.model_selection import StratifiedKFold, cross_val_score
from sklearn.neighbors import KNeighborsClassifier

valores_k_novo = list(range(1, 20))
medias_novo = []

cv_novo = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

for k_tmp_novo in valores_k_novo:
    modelo_tmp_novo = KNeighborsClassifier(n_neighbors=k_tmp_novo)
    scores_novo = cross_val_score(modelo_tmp_novo, entradas_treino_novo, classes_treino_novo, cv=cv_novo, scoring='accuracy')
    medias_novo.append(scores_novo.mean())

best_k_novo = valores_k_novo[int(np.argmax(medias_novo))]
print("Melhor K para o novo conjunto de features:", best_k_novo)

plt.figure(figsize=(6,4))
plt.plot(valores_k_novo, medias_novo, marker='o')
plt.title("Acurácia média (CV) por K (Novas Features)")
plt.xlabel("K")
plt.ylabel("Acurácia média (cv)")
plt.grid(True)
plt.show()

**Reasoning**:
Train a new KNN model using the best K found and evaluate its accuracy on the new test set. Then, generate a scatterplot to visualize the relationship between two of the new features, colored by diagnosis.



In [ ]:
# 4. Treine um novo modelo KNN
modelo_novo = KNeighborsClassifier(n_neighbors=best_k_novo)
modelo_novo.fit(entradas_treino_novo, classes_treino_novo)
print("Novo modelo treinado com K =", best_k_novo)

# 5. Avalie o modelo_novo no conjunto de teste
classes_encontradas_novo = modelo_novo.predict(entradas_teste_novo)
acertos_novo = accuracy_score(classes_teste_novo, classes_encontradas_novo)
print("Acerto médio de classificação com novas features: ", acertos_novo)

# 6. Gere um novo gráfico para visualizar as relações entre pelo menos duas das novas features
plt.figure(figsize=(8, 6))
sns.scatterplot(data=new_sub_df, x="textura_media", y="area_media", hue="diagnostico")
plt.title("Relação entre Textura Média e Área Média por Diagnóstico")
plt.xlabel("Textura Média")
plt.ylabel("Área Média")
plt.show()

**Reasoning**:
Compare the accuracy of the new model with the previous model and print a comment about the possible reasons for the difference. Then, finish the task.



In [ ]:
# 7. Compare a acurácia dos modelos
print("\nComparação de Acurácia:")
print(f"Acurácia do modelo anterior (features originais): {acertos:.4f}")
print(f"Acurácia do novo modelo (novas features): {acertos_novo:.4f}")

print("\nObservação:")
if acertos_novo > acertos:
    print("O novo modelo com as features selecionadas obteve uma acurácia maior.")
    print("Isso sugere que as novas features podem ser mais discriminativas para a classificação neste caso específico, ou que a combinação das novas features capturou melhor os padrões nos dados de treino.")
elif acertos_novo < acertos:
    print("O novo modelo com as features selecionadas obteve uma acurácia menor.")
    print("Isso sugere que as features originais selecionadas podem ter sido mais relevantes para a classificação, ou que a nova combinação de features não capturou tão bem os padrões necessários.")
else:
    print("Os dois modelos obtiveram acurácias semelhantes.")
    print("Isso sugere que as novas features selecionadas têm um poder discriminativo comparável ao conjunto de features originais para este modelo e conjunto de dados.")


## Summary:

### Data Analysis Key Findings

*   The loaded dataset contains 569 rows and 32 columns. The target variable `diagnostico` is imbalanced, with 357 benign cases (0) and 212 malignant cases (1).
*   Exploratory data analysis (EDA) revealed that features such as `raio_medio`, `concavidade_media`, and `pontos_concavos_medios` show distinct distributions between benign and malignant tumors, suggesting they are important for classification. Tumors with higher values for these features tend to be malignant.
*   The correlation heatmap indicated strong positive correlations between `diagnostico` and `raio_medio` (0.73), `concavidade_media` (0.70), and `pontos_concavos_medios` (0.78).
*   Cross-validation was used to select the optimal K value for the KNN model. For the original set of features, the best K found was 10.
*   The final KNN model, trained with K=10 on the original features, achieved an accuracy of approximately 90.35% on the unseen test set.
*   A practical exercise using a new subset of features (`textura_media`, `area_media`, etc.) resulted in a new best K of 7.
*   The new KNN model, trained with K=7 on the new feature subset, achieved an accuracy of approximately 86.84% on its respective test set.
*   Comparing the models, the one trained on the original set of features (accuracy $\approx$ 90.35%) performed slightly better than the one trained on the new feature subset (accuracy $\approx$ 86.84%).

### Insights or Next Steps

*   The original set of features (`raio_medio`, `concavidade_media`, `pontos_concavos_medios`) appears to be more discriminative for this classification task than the new subset tested.
*   Further analysis could involve exploring other feature selection techniques or testing different combinations of features to potentially improve model performance.
